src/lib/pci_virtio_net_vpnkit.c

```diff
@@ -324,21 +343,60 @@ static int vpnkit_connect(int fd, const char uuid[36], struct vif_info *vif)
 		init_reply.magic[4],
 		init_reply.version, (int)sizeof(init_reply.commit), init_reply.commit);
 
-	struct msg_ethernet cmd_ethernet = {
-		.command = CMD_ETHERNET,
-	};
-	memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
-
-	if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
-		return -1;
+	if (init_reply.version != init_msg.version) { 
+		fprintf(stderr, "virtio-net-vpnkit: protocol version mismatch: version %d requested, got version %d.\n", 
+			init_msg.version,
+			init_reply.version);
 	}
 
-	if (really_read(fd, (uint8_t*)vif, sizeof(*vif)) < 0) {
-		fprintf(stderr, "virtio-net-vpnkit: failed to read vif info\n");
+	if (preferred_ipv4 == NULL) { // ethernet
+		struct msg_ethernet cmd_ethernet = {
+			.command = CMD_ETHERNET,
+			.unused = 0,
+		};
+		memcpy(cmd_ethernet.uuid, uuid, sizeof(cmd_ethernet.uuid));
+
+		if (really_write(fd, (uint8_t*)&cmd_ethernet, sizeof(cmd_ethernet)) < 0) {
+			fprintf(stderr, "virtio-net-vpnkit: failed to write ethernet cmd\n");
+			return -1;
+		}
+
+	} else { // preferred_ipv4
+		struct msg_preferred_ipv4 cmd_preferred_ipv4 = {
+			.command = CMD_PREFERRED_IPV4,
+			.ip = OSSwapInt32(*preferred_ipv4), // VPNKit uses LE, our IP is in network byte order
```

Comment:
Is VPNKit explicitly in LE or is it in host byte order. If the latter this should be `htonl` or equiv

Reply:
VPNKit explicitly uses LE